File: app/src/main/java/fr/skyost/timetable/widget/TodayWidgetUpdateScheduler.java
package fr.skyost.timetable.widget;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;

import org.joda.time.DateTime;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import fr.skyost.timetable.lesson.Lesson;
import fr.skyost.timetable.lesson.database.LessonDao;
import fr.skyost.timetable.utils.Utils;

/**
 * The AsyncTask that allows to schedule a widget update.
 */

public class TodayWidgetUpdateScheduler extends AsyncTask<LessonDao, Void, DateTime> {

	/**
	 * A context atomic reference.
	 */

	private final AtomicReference<Context> context;

	/**
	 * Creates a new today's widget update scheduler task.
	 *
	 * @param context The context.
	 */

	TodayWidgetUpdateScheduler(final Context context) {
		this.context = new AtomicReference<>(context);
	}

	@Override
	protected DateTime doInBackground(final LessonDao... daos) {
		// We get the DAO.
		final LessonDao dao = daos[0];

		// We get the remaining lessons and if possible, we return the end of the next one.
		final List<Lesson> remainingLessons = dao.getRemainingLessons();
		final DateTime date = remainingLessons.isEmpty() ? Utils.tomorrowMidnight() : remainingLessons.get(0).getEndDate();
		return date.getSecondOfMinute() == 0 ? date.withSecondOfMinute(1) : date;
	}

	@Override
	protected void onPostExecute(final DateTime date) {
		final Context context = this.context.get();
		if(context == null) {
			return;
		}

		// With the alarm manager, we schedule our update.
		final AlarmManager manager = (AlarmManager)context.getSystemService(Context.ALARM_SERVICE);
		if(manager == null) {
			return;
		}

		final Intent intent = new Intent(context, getClass());
		intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
		intent.putExtra(TodayWidgetReceiver.INTENT_REFRESH_WIDGETS, true);

		manager.set(AlarmManager.RTC_WAKEUP, date.getMillis(), PendingIntent.getBroadcast(context, TodayWidgetReceiver.SCHEDULE_REQUEST, intent, PendingIntent.FLAG_UPDATE_CURRENT));
	}

	/**
	 * Returns the context reference.
	 *
	 * @return The context reference.
	 */

	public AtomicReference<Context> getContext() {
		return context;
	}

}

File: android/app/src/main/kotlin/fr/skyost/timetable/widget/TodayWidgetUpdateScheduler.kt
package fr.skyost.timetable.widget

import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import fr.skyost.timetable.Lesson
import java.time.LocalDateTime
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.concurrent.Executors

/**
 * The utility class that allows to schedule a widget update.
 */
class TodayWidgetUpdateScheduler {
    companion object {
        fun schedule(context: Context, widgetId: Int, lessonsOfTheDay: List<Lesson>) {
            val executor = Executors.newSingleThreadExecutor()
            val handler = Handler(Looper.getMainLooper())

            executor.execute {
                val nextSchedule = getNextSchedule(lessonsOfTheDay)
                handler.post {
                    scheduleAt(context, widgetId, nextSchedule)
                }
            }
        }

        private fun getNextSchedule(lessonsOfTheDay: List<Lesson>): LocalDateTime {
            val tomorrowMidnight = LocalDateTime
                .now()
                .plusDays(1)
                .truncatedTo(ChronoUnit.DAYS)
            try {
                // We get the remaining lessons and if possible, we return the end of the next one.
                val remainingLessons: List<Lesson> = getRemainingLessons(lessonsOfTheDay)
                val date: LocalDateTime =
                    if (remainingLessons.isEmpty()) tomorrowMidnight else remainingLessons[0].end
                return if (date.second == 0) date.withSecond(1) else date
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
            return tomorrowMidnight
        }

        private fun getRemainingLessons(lessonsOfTheDay: List<Lesson>): List<Lesson> {
            // We get the today's lessons.
            val now: LocalDateTime = LocalDateTime.now()
            val result = ArrayList(lessonsOfTheDay)
            for (lesson in lessonsOfTheDay) {
                // If we are past the lesson, we remove it from the list.
                if (!now.isAfter(lesson.end)) {
                    continue
                }
                result.remove(lesson)
            }
            return result
        }

        private fun scheduleAt(context: Context, widgetId: Int, date: LocalDateTime) {
            // With the alarm manager, we schedule our update.
            val offset = ZonedDateTime.now().offset
            val manager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            val intent = Intent(context, TodayWidgetProvider::class.java)
            intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
            intent.putExtra(TodayWidgetProvider.INTENT_REFRESH_WIDGETS, true)
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
            manager[AlarmManager.RTC_WAKEUP, date.toEpochSecond(offset) * 1000] =
                PendingIntent.getBroadcast(
                    context,
                    TodayWidgetProvider.REFRESH_REQUEST,
                    intent,
                    TodayWidgetProvider.FLAG_IMMUTABLE_OR_UPDATE_CURRENT
                )
        }
    }
}
